### ISSUE TITLE
hope add terminationGracePeriodSeconds support

### ISSUE THREAD

**Issue by lurenjia528** [prometheus-operator/prometheus-operator#3433]
**What is missing?**
hope add `sts.spec.template.spec.terminationGracePeriodSeconds`  support

**Why do we need it?**
releate  to　`https://github.com/prometheus-operator/prometheus-operator/issues/3425` ,

if prometheus(sts) is running on node-1,when node-1 is down, the prometheus(sts) do not recreate on other ready node,
yeah,sts is design  so , https://github.com/kubernetes/kubernetes/issues/54368#issuecomment-339378164 ,  but when set terminationGracePeriodSeconds=0, sts can recreate on other ready node,then ,this enables fault migration 

so ,I hope add terminationGracePeriodSeconds in [prometheusSpec](https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/api.md#prometheusspec)

**Comment by brancz** [comment]
I agree with @coderanger that duplicating is in fact the correct strategy here, I don't want to expose a field that enables an anti-pattern.

**Comment by simonpasquier** [comment]
Irrespective of the use case mentioned by @lurenjia528 (immediate shutdown), there might be a need to customize the `TerminationGracePeriodSeconds` to a higher value than the default 600s?
We've got a [report](https://bugzilla.redhat.com/show_bug.cgi?id=1883749) about Prometheus containers requiring close to 10 minutes to flush the head to WAL on shutdown (not sure exactly why it took so long, maybe the node was overloaded). In one occasion, the container was presumably SIGKILLed and the data on disk was corrupted after that.

**Comment by brancz** [comment]
@simonpasquier I think we should focus on the Prometheus improvements that are happening with checkpoints and snapshots on shutdown, so that won't be necessary.

**Comment by simonpasquier** [comment]
#3673 is another report about the hard-coded termination grace period. This looks similar to #3278: some cloud providers enforce a maximum value for `terminationGracePeriodSeconds` preventing the roll-out of the prometheus statefulset.

**Comment by diyavicky** [comment]
@simonpasquier So at this point there is no possible way to provide acceptable custom value for terminationGracePeriod?
Just trying to understand the current state of this issue.

Thanks!

**Comment by diyavicky** [comment]
@simonpasquier @brancz Could you please advise if there is a possible way to mention custom terminationGracePeriod by any chance?

Please advise!

Thanks!

**Comment by diyavicky** [comment]
@simonpasquier @brancz Really appreciate any update on the above query.

Thanks!

**Comment by brancz** [comment]
You can always merge arbitrary customization to containers by specifying a container with the name "prometheus" in the "containers" field of the Prometheus custom resource, note this is an unsupported feature and customizations you do we won't help troubleshoot. We are unlikely to accept a first class feature for customizing `terminationGracePeriod`, as we would prefer to fix the root cause.

**Comment by aa-belikov** [comment]
> You can always merge arbitrary customization to containers by specifying a container with the name "prometheus" in the "containers" field of the Prometheus custom resource, note this is an unsupported feature and customizations you do we won't help troubleshoot. We are unlikely to accept a first class feature for customizing `terminationGracePeriod`, as we would prefer to fix the root cause.

terminationGracePeriodSeconds is a pod level parameter.
Is there another way to change that?

**Comment by diranged** [comment]
So.. just to double check, there is still no answer here right? Even though we need to specifically set the `terminationGracePeriod` (per https://ganeshvernekar.com/blog/prometheus-tsdb-snapshot-on-shutdown/#few-things-to-be-aware-of) to make sure that prometheus can shut down gracefully?

**Comment by camclay** [comment]
+1 some of our teams have noticed issues with Spot instances, as there was a change in December that hardcoded the termination to 600 seconds -> https://github.com/prometheus-operator/prometheus-operator/commit/43b66853bddf42dcf702140a2222c7683773768a 

Proms on spot instances, have a max of 120 seconds to shutdown, and this now ignores those shortened requests.

**Comment by simonpasquier** [comment]
https://github.com/prometheus-operator/prometheus-operator/commit/43b66853bddf42dcf702140a2222c7683773768a didn't change the default value which has been 600s for a while now. Can you provide more details on the spot instance issue? I fail to understand exactly the problem.
